ocaml/xapi/xha_statefile.ml

```diff
@@ -89,15 +89,54 @@ let assert_sr_can_host_statefile ~__context ~sr ~cluster_stack =
              ]
            )
         )
-  | (_, sm) :: _ ->
+  | (_, sm) :: _ -> (
       if
         (not (List.mem_assoc "VDI_GENERATE_CONFIG" sm.Db_actions.sM_features))
         && not (List.mem_assoc "VDI_ATTACH_OFFLINE" sm.Db_actions.sM_features)
       then
         raise
           (Api_errors.Server_error
              (Api_errors.sr_operation_not_supported, [Ref.string_of sr])
+          ) ;
+      let size = minimum_size in
+      match
+        List.filter
+          (fun self ->
+            true
+            && Db.VDI.get_type ~__context ~self = `ha_statefile
+            && Db.VDI.get_virtual_size ~__context ~self >= size
           )
+          (Db.SR.get_VDIs ~__context ~self:sr)
+      with
+      | x :: _ ->
+          debug "Would re-use existing statefile: %s"
+            (Db.VDI.get_uuid ~__context ~self:x) ;
+          Some x
+      | [] ->
+          debug
+            "no suitable existing statefile found; would have to create a \
+             fresh one" ;
+          let free_space =
+            Int64.sub
+              (Db.SR.get_physical_size ~__context ~self:sr)
+              (Db.SR.get_physical_utilisation ~__context ~self:sr)
+          in
+          if free_space < minimum_size then (
+            info "SR %s has %Ld free space, needed %Ld" (Ref.string_of sr)
+              free_space minimum_size ;
+            raise
+              (Api_errors.Server_error
+                 (Api_errors.sr_source_space_insufficient, [Ref.string_of sr])
+              )
+          ) else
+            None
+    )
+
+let assert_sr_can_host_statefile ~__context ~sr ~cluster_stack =
+  let (_ : 'a option) =
+    check_sr_can_host_statefile ~__context ~sr ~cluster_stack
```

Comment:
Might be worth using Result.t instead of exceptions

Reply:
exceptions are needed because this function is eventually exposed in the API, which has to return an error. We can clean this up, but that'd need a lot more changes inside the 'check_sr_can_host_statefile' function to convert all the exceptions in to results, only for at the end to convert the results back into exceptions.